### PR TITLE
Add "provisional" and "internal" marker support for matter IDL files

### PR DESCRIPTION
**Pull request by andy31415**
The intent is to eventually track what APIs are allowed to be changed in a backwards-incompatible way (i.e. remove fields or rename names).

This assumes that we will have this info in some way in zap xml and later in CSA xml.

### CODE REVIEW

File: scripts/py_matter_idl/matter_idl/matter_grammar.lark [hunk 1]
```diff
@@ -2,6 +2,13 @@ struct: struct_qualities "struct"i id "{" (struct_field ";")* "}"
 struct_quality: "fabric_scoped"i -> struct_fabric_scoped
 struct_qualities: struct_quality*
 
+// Composing elements may be at different maturity level
+// This only contains non-stable items. Items without a maturity
+// flag are to be assumed "stable"
```
Comment (tcarmelveilleux) [inline scripts/py_matter_idl/matter_idl/matter_grammar.lark:7]: It should be optionally possible to explicitly state "stable"

**Review by tcarmelveilleux — APPROVED** [review]

File: scripts/py_matter_idl/matter_idl/matter_grammar.lark (hunk 1 above)
Comment (andy31415) [reply scripts/py_matter_idl/matter_idl/matter_grammar.lark:7]: Added as a supported keyword. 

I would however generally discourage its use.

File: scripts/py_matter_idl/matter_idl/matter_grammar.lark [hunk 2]
```diff
@@ -78,13 +87,13 @@ bool_default: "true"i  -> bool_default_true
             | "false"i -> bool_default_false
 ?default_value: "default"i "=" (integer | ESCAPED_STRING | bool_default)
 
-constant_entry: id "=" positive_integer ";"
+constant_entry: [maturity] id "=" positive_integer ";"
```
Comment (bzbarsky-apple) [inline scripts/py_matter_idl/matter_idl/matter_grammar.lark:90]: I assume this covers enum and bitmap values?

File: scripts/py_matter_idl/matter_idl/matter_grammar.lark [hunk 3]
```diff
@@ -78,13 +87,13 @@ bool_default: "true"i  -> bool_default_true
...
 positive_integer: POSITIVE_INTEGER | HEX_INTEGER
 negative_integer: "-" positive_integer
 
 integer: positive_integer | negative_integer
 
-struct_field: member_attribute* field
+struct_field: [maturity] member_attribute* field
```
Comment (bzbarsky-apple) [inline scripts/py_matter_idl/matter_idl/matter_grammar.lark:96]: What about command fields?  Event fields?  Or do they all use this production?

File: scripts/py_matter_idl/matter_idl/matter_grammar.lark (hunk 2 above)
Comment (andy31415) [reply scripts/py_matter_idl/matter_idl/matter_grammar.lark:90]: yes, both contain constants.

File: scripts/py_matter_idl/matter_idl/matter_grammar.lark (hunk 3 above)
Comment (andy31415) [reply scripts/py_matter_idl/matter_idl/matter_grammar.lark:96]: structs apply to all commands (they are structs for both in and out) as well as events (they use struct field too).

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**